src/prototype.rs

```diff
@@ -166,11 +167,65 @@ fn first_pass(s: &str) -> Tree<Item> {
             // blank line
             ix += scan_eol(&s[ix..]).0;
         } else {
-            let (mut leading_bytes, mut leading_space) = scan_leading_space(&s[ix..], 0);
+            let (leading_bytes, leading_spaces) = scan_leading_space(&s[ix..], 0);
             ix += leading_bytes;
 
+            let (atx_size, atx_level) = scan_atx_header(&s[ix..]);
+            if atx_level > 0 && leading_spaces < 4 {
+                // println!("Found atx");
+                tree.append(Item {
+                    start: ix,
+                    end: 0, // set later
+                    body: ItemBody::Header(atx_level),
+                });
+                ix += atx_size;
+                // next char is space or scan_eol
+                // (guaranteed by scan_atx_header)
+                let b = s.as_bytes()[ix];
+                if b == b'\n' || b == b'\r' {
+                    ix += scan_eol(&s[ix..]).0;
+                    continue;
+                }
+                // skip leading spaces
+                let skip_spaces = scan_ch_repeat(&s[ix..], b' ');
+                ix += skip_spaces;
+
+                // now handle the header text
+                let header_start = ix;
+                let header_node_idx = tree.cur; // so that we can set the endpoint later
+                tree.push();
+                let header_text_size = parse_line(&mut tree, s, ix);
+                ix += header_text_size;
+                tree.nodes[header_node_idx].item.end = ix;
+
+
+                // remove trailing matter from header text
+                let header_text = &s[header_start..];
+                let mut limit = ix - header_start;
+                while limit > 0 && header_text.as_bytes()[limit-1] == b' ' {
```

Comment:
There are two alternatives to writing out these loops: [rposition](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.rposition) and [memchrr](https://docs.rs/memchr/1.0.1/memchr/fn.memrchr.html). In this case I'm not sure I'd change it, because the expected number of characters is small, but good to be aware of the choices.